crates/nu-protocol/src/value/filesize.rs

```diff
@@ -432,97 +526,87 @@ impl fmt::Display for FilesizeUnit {
-            "kb" | "kib" => Some(byte_unit::Unit::KiB),
-            "mb" | "mib" => Some(byte_unit::Unit::MiB),
-            "gb" | "gib" => Some(byte_unit::Unit::GiB),
-            "tb" | "tib" => Some(byte_unit::Unit::TiB),
-            "pb" | "pib" => Some(byte_unit::Unit::TiB),
-            "eb" | "eib" => Some(byte_unit::Unit::EiB),
-            _ => None,
+impl fmt::Display for DisplayFilesize {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self {
+            filesize,
+            unit,
+            precision,
+        } = *self;
+        let precision = precision.or(f.precision());
+        match unit {
+            FilesizeUnit::B => write!(f, "{} B", filesize.0),
+            FilesizeUnit::KiB
+            | FilesizeUnit::MiB
+            | FilesizeUnit::GiB
+            | FilesizeUnit::TiB
+            | FilesizeUnit::PiB
+            | FilesizeUnit::EiB => {
+                let val = filesize.0 as f64 / unit.as_bytes() as f64;
+                // This won't give exact results for large filesizes and/or units.
+                if let Some(precision) = precision {
+                    write!(f, "{val:.precision$} {unit}")
+                } else {
+                    write!(f, "{val} {unit}")
+                }
+            }
+            FilesizeUnit::KB
+            | FilesizeUnit::GB
+            | FilesizeUnit::MB
+            | FilesizeUnit::TB
+            | FilesizeUnit::PB
+            | FilesizeUnit::EB => {
+                // Format an exact, possibly fractional, string representation of `filesize`.
+                let bytes = unit.as_bytes() as i64;
+                let whole = filesize.0 / bytes;
+                let fract = (filesize.0 % bytes).unsigned_abs();
+                if precision == Some(0) || (precision.is_none() && fract == 0) {
+                    write!(f, "{whole} {unit}")
+                } else {
+                    // fract <= bytes by nature of `%` and bytes <= EB = 10 ^ 18
+                    // So, the longest string for the fractional portion can be 18 characters.
+                    let buf = &mut [b'0'; 18];
+                    let stop = precision.unwrap_or(buf.len());
+                    let mut fract = fract;
+                    let mut power = bytes.unsigned_abs() / 10;
+                    let mut i = 0;
+                    loop {
+                        let q = fract / power;
+                        let r = fract % power;
+                        debug_assert!(q < 10);
+                        buf[i] += q as u8;
+                        i += 1;
+                        if r == 0 || i >= stop {
+                            break;
+                        }
+                        fract = r;
+                        power /= 10;
+                    }
+
+                    // Safety: all the characters in `buf` are valid UTF-8.
+                    let fract = unsafe { std::str::from_utf8_unchecked(&buf[..i]) };
```

Comment:
The `debug_assert` above should catch potential errors that could make this invalid UTF-8. I could change this to `std::str::from_utf8` paired with an `expect` instead if that is better.